runtime/compare.c

```diff
@@ -118,9 +136,16 @@ static intnat do_compare_val(struct compare_stack* stk,
 {
   struct compare_item * sp;
   tag_t t1, t2;
+  int signal_poll_timer = 0;
 
   sp = stk->stack;
   while (1) {
+    /* Periodically poll for actions, since this loop can run for
+       unbounded time. */
+    if (CAMLunlikely(++signal_poll_timer >= COMPARE_SIGNAL_POLL_PERIOD)) {
+      signal_poll_timer = 0;
+      poll_pending_actions(stk);
```

Comment:
It seems like `v1` and `v2` are kept live across this without being registered as roots.

Reply:
Right... I moved the poll logic to take place just before the `next_item` label, at this point `v1, v2` are not kept live as they will be reloaded from the compare stack.

Reply:
(It would also be possible to keep the poll test at the top and root v1, v2 at this point, but doing so requires to copy them into an array to not pessimize the code.)

Reply:
Another option would be to pass the addresses of v1 and v2 and treat them as arrays of size one for `Begin_roots_block`.

Reply:
Right, the compiler cannot optimize the spilling/reloading as initially thought. However, by manually copying/restoring v1 and v2 within the inner block, and rooting the copies as suggested, one gets a performance within the irrelevant code layout effects. Strangely, copying into an array instead of two variables and rooting the array as you did is not equivalent, it comes with a huge cost.

That being said, I got accustomed to the more complicated version, so feel free to choose what works best for you.